src/Tokenomia/Vesting/PrivateSale.hs

```diff
@@ -0,0 +1,223 @@
+    ) =>
+    PrivateSale ->
+    [B.TxHash] ->
+    m ()
+verifyTxs ps =
+    let invs = getPsInvestments ps
+     in mapM_ (verifyTx invs)
+  where
+    verifyTx ::
+        ( MonadIO m
+        , MonadError TokenomiaError m
+        , MonadReader Environment m
+        ) =>
+        [Investment] ->
+        B.TxHash ->
+        m ()
+    verifyTx invs txh = do
+        bfTxUtxos <- getTxUtxosByTxHash txh -- :: B.TransactionUtxos
+        inv <- liftEither $ getInvByTxHash txh invs -- :: Investment
+        let bfUtxoOutputs = bfTxUtxos ^. B.outputs -- :: [B.UtxoOutput]
+            treasAddrOutputs = filter (\output -> (output ^. B.address) == (ps ^. psAddress)) bfUtxoOutputs -- :: [B.UtxoOutput]
+            bfAmts = concat ((^. B.amount) <$> treasAddrOutputs) -- :: [B.Amount]
+            bfValues = amountToAssetValue <$> bfAmts -- :: [(Value.AssetClass, Integer)]
+            confirmedVals = confirmValues ps (traceShowId bfValues) -- :: Integer
+            invAmount' = inv ^. invAmount -- :: Integer
+            invAssetClass' = inv ^. invAssetClass -- :: Value.AssetClass
+         in unless (invAssetClass' == ps ^. psAssetClass && confirmedVals == invAmount') $
+                throwError . BlockFrostError . B.BlockfrostError $ "Values don't match"
+
+confirmValues :: PrivateSale -> [(Value.AssetClass, Integer)] -> Integer
+confirmValues ps = foldr (\val z -> if fst val == ps ^. psAssetClass then snd val + z else z) 0
+
+getTxUtxosByTxHash ::
+    ( MonadIO m
+    , MonadError TokenomiaError m
+    , MonadReader Environment m
+    ) =>
+    B.TxHash ->
+    m B.TransactionUtxos
+getTxUtxosByTxHash txh = do
+    prj <- projectFromEnv''
+    eitherErrTxUtxos <- liftIO $ B.runBlockfrost prj (B.getTxUtxos txh)
+    liftEither $ first BlockFrostError eitherErrTxUtxos
+
+getInvByTxHash :: B.TxHash -> [Investment] -> Either TokenomiaError Investment
+getInvByTxHash txh invs =
+    maybe
+        ( Left . BlockFrostError . B.BlockfrostError $
+            "Investment list doesn't contain matching TxHash"
+        )
+        Right
+        (find (\inv -> (inv ^. invTx) == txh) invs)
+
+getPsInvestments ::
+    PrivateSale ->
+    [Investment]
+getPsInvestments ps = investments
+  where
+    investments :: [Investment]
+    investments = concat ((^. piInvestments) <$> investors)
+
+    investors :: [PrivateInvestor]
+    investors = ps ^. psInvestors
+
+verifyTxHashList :: [B.TxHash] -> [B.AddressTransaction] -> [B.TxHash]
+verifyTxHashList txhs addrTxs =
+    filter (`elem` txhs) ((^. B.txHash) <$> addrTxs)
+
+amountToAssetValue :: B.Amount -> (Value.AssetClass, Integer)
+amountToAssetValue (B.AdaAmount ll) = (Value.assetClass "" "", someDiscreteAmount $ toSomeDiscrete ll)
+amountToAssetValue (B.AssetAmount sd) = (Value.assetClass (fromString cs) (fromString name), someDiscreteAmount sd)
+  where
+    encodedName = unpack $ someDiscreteCurrency sd
+    cs = take 56 encodedName
+    name = either error id $ unhex $ drop 56 encodedName
```

Comment:
This function is now partial, which I don't think is ideal. I'd say it's worth it to rework it to return an `Either _ (Value.AssetClass, Integer)`

Reply:
Unsure if this error case is worth trying to catch, as if it does occur, it means blockfrost is broken, and the script should fail out in a more "angry" fashion. We could introduce the `MonadError` constraint and add a new type, or reuse an existing blockfrost type, may not be worth it though.

Reply:
Agreed, lets do that

Reply:
So just adding a comment explaining this, not adding `MonadError` or anything right?